cylc/flow/simulation.py

```diff
@@ -0,0 +1,305 @@
+
+    This is for Dummy mode only.
+    """
+    script = "sleep %d" % sleep_sec
+    # Dummy message outputs.
+    for msg in rtc['outputs'].values():
+        script += "\ncylc message '%s'" % msg
+    if rtc['simulation']['fail try 1 only']:
+        arg1 = "true"
+    else:
+        arg1 = "false"
+    arg2 = " ".join(rtc['simulation']['fail cycle points'])
+    script += "\ncylc__job__dummy_result %s %s || exit 1" % (arg1, arg2)
+    return script
+
+
+def disable_platforms(
+    rtc: Dict[str, Any]
+) -> None:
+    """Force platform = localhost
+
+    Remove legacy sections [job] and [remote], which would conflict
+    with setting platforms.
+
+    TODO (Cylc 9): This function can probably be removed
+    when the items in FORBIDDEN_WITH_PLATFORM with platforms
+    are removed from the config - it will become trivially simple.
+    """
+    for section, keys in FORBIDDEN_WITH_PLATFORM.items():
+        if section in rtc:
+            for key in keys:
+                if key in rtc[section]:
+                    rtc[section][key] = None
+    rtc['platform'] = 'localhost'
+
+
+def parse_fail_cycle_points(
+    f_pts_orig: list
+) -> Union[None, List[PointBase]]:
+    """Parse `[simulation][fail cycle points]`.
+
+    - None for "fail all points".
+    - Else a list of cycle point objects.
+
+    Examples:
+        >>> this = parse_fail_cycle_points
+        >>> this(['all']) is None
+        True
+        >>> this([])
+        []
+    """
+    f_pts: Optional[List[PointBase]]
+    if f_pts_orig is None or 'all' in f_pts_orig:
+        f_pts = None
+    else:
+        f_pts = []
+        for point_str in f_pts_orig:
+            f_pts.append(get_point(point_str).standardise())
+    return f_pts
+
+
+def unpack_dict(dict_: NestedDict, parent_key: str = '') -> Dict[str, Any]:
+    """Unpack a nested dict into a single layer.
+
+    Examples:
+        >>> unpack_dict({'foo': 1, 'bar': {'baz': 2, 'qux':3}})
+        {'foo': 1, 'bar.baz': 2, 'bar.qux': 3}
+        >>> unpack_dict({'foo': {'example': 42}, 'bar': {"1":2, "3":4}})
+        {'foo.example': 42, 'bar.1': 2, 'bar.3': 4}
+
+    """
+    output = {}
+    for key, value in dict_.items():
+        new_key = parent_key + '.' + key if parent_key else key
+        if isinstance(value, dict):
+            output.update(unpack_dict(value, new_key))
+        else:
+            output[new_key] = value
+
+    return output
+
+
+def nested_dict_path_update(
+    dict_: NestedDict, path: List[Any], value: Any
+) -> NestedDict:
+    """Set a value in a nested dict.
+
+    Examples:
+        >>> nested_dict_path_update({'foo': {'bar': 1}}, ['foo', 'bar'], 42)
+        {'foo': {'bar': 42}}
+    """
+    this = dict_
+    for i in range(len(path)):
+        if isinstance(this[path[i]], dict):
+            this = this[path[i]]
+        else:
+            this[path[i]] = value
+    return dict_
+
+
+def update_nested_dict(rtc: NestedDict, dict_: NestedDict) -> None:
```

Comment:
This and the two functions above feel like I've replicated functionality from somewhere else (parsec?, OrderedDictWithDefaults?), but I cannot for the life of me work out how to use existing functions.